crates/typst-pdf/src/color.rs

```diff
@@ -377,26 +378,39 @@ impl PaintEncode for Color {
 /// Extra color space functions.
 pub(super) trait ColorSpaceExt {
     /// Returns the range of the color space.
-    fn range(self) -> [f32; 6];
+    fn range(self) -> SmallVec<[f32; 8]>;
 
     /// Converts a color to the color space.
-    fn convert<U: QuantizedColor>(self, color: Color) -> [U; 3];
+    fn convert<U: QuantizedColor>(self, color: Color) -> SmallVec<[U; 4]>;
 }
 
 impl ColorSpaceExt for ColorSpace {
-    fn range(self) -> [f32; 6] {
-        [0.0, 1.0, 0.0, 1.0, 0.0, 1.0]
+    fn range(self) -> SmallVec<[f32; 8]> {
+        match self {
+            ColorSpace::Cmyk => smallvec![0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0],
+            ColorSpace::D65Gray => smallvec![0.0, 1.0],
+            _ => smallvec![0.0, 1.0, 0.0, 1.0, 0.0, 1.0],
+        }
     }
 
-    fn convert<U: QuantizedColor>(self, color: Color) -> [U; 3] {
+    fn convert<U: QuantizedColor>(self, color: Color) -> SmallVec<[U; 4]> {
         let range = self.range();
-        let [x, y, z, _] = self.encode(color);
+        let [x, y, z, a] = self.encode(color);
 
-        [
-            U::quantize(x, [range[0], range[1]]),
-            U::quantize(y, [range[2], range[3]]),
-            U::quantize(z, [range[4], range[5]]),
-        ]
+        match self {
+            ColorSpace::Cmyk => smallvec![
+                U::quantize(x, [range[0], range[1]]),
+                U::quantize(y, [range[2], range[3]]),
+                U::quantize(z, [range[4], range[5]]),
+                U::quantize(a, [range[6], range[7]]),
+            ],
+            ColorSpace::D65Gray => smallvec![U::quantize(x, [range[0], range[1]])],
+            _ => smallvec![
+                U::quantize(x, [range[0], range[1]]),
+                U::quantize(y, [range[2], range[3]]),
+                U::quantize(z, [range[4], range[5]]),
+            ],
+        }
```

Comment:
Maybe you could use the result from the `range` function above to generalize this:
```rs
let ranges = self.range().chunks(2);
let components = self.encode(color);

ranges
    .zip(components)
    .map(|(range, component)| U::quantize(component, [range[0], range[1]]))
    .collect()
```

Reply:
Quite neat! Thanks for the suggestion!